pyccel/ast/object_address.py

```diff
@@ -0,0 +1,38 @@
+#------------------------------------------------------------------------------------------#
+# This file is part of Pyccel which is released under MIT License. See the LICENSE file or #
+# go to https://github.com/pyccel/pyccel/blob/master/LICENSE for full license details.     #
+#------------------------------------------------------------------------------------------#
+
+"""
+Module representing object address.
+"""
+
+from .basic import PyccelAstNode
+
+from .variable import Variable, IndexedElement
+
+class ObjectAddress(PyccelAstNode):
+    """Represents the address of an object.
+    ObjectAddress(Variable('int','a'))                            is  &a
+    ObjectAddress(Variable('int','a', memory_handling='alias'))   is   a
+    """
+
+    __slots__ = ('_variable', '_rank', '_precision', '_dtype', '_shape', '_order')
+    _attribute_nodes = ('_variable',)
+
+    def __init__(self, variable):
+        if not isinstance(variable, (Variable, IndexedElement, ObjectAddress)):
+            raise TypeError("object must be a Variable, IndexedElement or ObjectAddress.") #TODO: to change
```

Comment:
Shouldn't this PR remove this restriction?

Reply:
Do you think it's a good idea to remove the attributes of `ObjectAddress`? because somewhere in the code we access the attributes `rank` or `order` for example, even though the target object doesn't have these attributes.
Like here: https://github.com/pyccel/pyccel/blob/6eb29eb0938354e1b65281c290464a1461089c84/pyccel/codegen/printing/ccode.py#L1656

Reply:
I think this restriction should be replaced with :
```
if not isinstance(var, PyccelAstNode):
```
`PyccelAstNode`s always have these attributes and it doesn't make sense to take the address of something that is not a `PyccelAstNode` (eg an `If`)

Reply:
Did you try removing the `try`/`except` mentioned above to see if anything breaks?

Reply:
The same exception is raised

Reply:
it was `PyBuildValueNode`

Reply:
Hmm, this class probably ought to inherit from `PyccelAstNode`. It is basically a `FunctionCall` and that class does inherit from `PyccelAstNode`. The properties are then set to match the results of the function call, so for `PyBuildValueNode` that would be:
- dtype : `PyccelPyObject`
- precison : 0
- rank : 0
- shape : ()
- order : None

So then to make it work we would have to handle this class in `stored_in_c_pointer` (should return true)

Reply:
Please could you make this change before merging. The handling in `stored_in_c_pointer` should be done in the `cwrappercode.py` file

Reply:
By "this chang" I also meant making it inherit from `PyccelAstNode`.

Reply:
The changes to be made are therefore:
- Make `PyBuildValueNode` inherit from `PyccelAstNode` (set class variables as described, add `_attribute_nodes` class property, don't forget to call superclass's `__init__` method)
- Correct `stored_in_c_pointer` to recognise this object as a pointer
- Remove the try/except clause

Reply:
Same problem for `Nil`, I tried to apply same solution as for `PyBuildValueNode` but was causing a cyclic imports.

Reply:
Cyclic imports via what? What imports did you need to add? Doesn't `Nil` inherit from `Literal`?

Reply:
I've just checked the code. I'm still not sure how you managed to create a cyclic import but I see that `Nil` currently inherits from `Basic` and is a singleton. If all the `PyccelAstNode` attributes are class properties then I don't think that should create problems, but if it does then maybe this special class should be handled explicitly by `ObjectAddress`

Reply:
Ah. That makes sense. I'm not sure what the dtype should be here, possibly `NativeGeneric`. It definitely shouldn't be `PyccelPyObject` though as null is equivalent to C's `NULL` pointer which can be used in many more places than a [PyObject](https://pythonextensionpatterns.readthedocs.io/en/latest/refcount.html#pyobjects-and-reference-counting)

Reply:
The possible dtypes can be found [here](https://github.com/pyccel/pyccel/blob/master/pyccel/ast/datatypes.py)

I would suggest one of:
- `NativeGeneric` - means any type
- `NativeVoid`

Reply:
If you come across other types that need this treatment try and work out what the dtype should be. For example [PyBuildValueNode](http://web.mit.edu/people/amliu/vrut/python/ext/buildValue.html) has the dtype `PyccelPyObject` because it represents a call to the following function:
```c
PyObject *Py_BuildValue(char *format, ...);
```

For a function call the dtype matches the result type

Reply:
fixed in https://github.com/pyccel/pyccel/pull/1151/commits/da63217bab5688b37c467f694024648c5f0e082c https://github.com/pyccel/pyccel/pull/1151/commits/7adb2d049247787cea7b17cecddc0de2718bf1f2